### PR TITLE
fix(deploy): make rollout-gate suppression of machine-config sync visible and bounded

### LINKED ISSUE #2951: Cilium rollout gate silently suppresses Talos machine-config sync indefinitely
> 🤖 Generated by the Agentic Engineer

## Evidence

The `🔄 Update cluster (Talos machine config)` step in `.github/actions/deploy-prod/action.yml` — the platform's **only** Talos machine-config sync — has been skipped on every prod deploy for at least four days, while each deploy reported success.

| Deploy | `__self.cluster_update` |
|---|---|
| 2026-08-04T16:09Z ([30927795575](https://github.com/devantler-tech/platform/actions/runs/30927795575)) | `outcome=skipped duration_ms=0` |
| 2026-08-04T15:33Z ([30924710728](https://github.com/devantler-tech/platform/actions/runs/30924710728)) | `outcome=skipped` |
| 2026-08-03T07:58Z ([30795581205](https://github.com/devantler-tech/platform/actions/runs/30795581205)) | `outcome=skipped` |
| 2026-08-01T11:04Z ([30696960482](https://github.com/devantler-tech/platform/actions/runs/30696960482)) | `outcome=skipped` |

The cause is the first arm of its condition:

```yaml
if: >-
  steps.cilium_rollout_gate.outputs.active != 'true' &&
  steps.wait_flux_revision.outcome == 'success'
```

In the 16:09Z run `wait_flux_revision` was `success`, and the gate step printed `Cilium homogeneous-device rollout gate active: Cluster Autoscaler is suspended.` The gate is still active: `kube-system/cluster-autoscaler-hetzner-cluster-autoscaler` reads **`0/0`** on prod.

## The problem

The gate is deliberate and its reasoning is sound — KSail owns Cluster Autoscaler and `cluster update` could reconcile it back to one replica mid-rollout, racing a scale-up before the first Cilium canary is verified. The step's own comment describes it as **temporary**, released by *"a reviewed rollout-completion or rollback artifact"*.

The defect is not the coupling, it is that the coupling is **unbounded and silent**:

- A safety gate for *one* component indefinitely disables an *unrelated* reconcile loop.
- Nothing reports it. The deploy is green, the summary says success, and the skipped step reads as normal conditional behaviour.
- The longer it holds, the larger the un-reconciled machine-config delta becomes — and the first `cluster update` after release applies all of it at once, against nodes that have drifted.

It also silently blocks unrelated work: [#2938](https://github.com/devantler-tech/platform/issues/2938) and [#2922](https://github.com/devantler-tech/platform/issues/2922) both depend on `cluster update` applying the root `OCIRepository` cosign `verify` block. KSail shipped that fix in v7.178.13 and this repo already pins v7.178.14, so both issues look like KSail defects and are in fact blocked here.

## Affected audience and impact

Operators of the prod cluster. Machine config in Git is not the machine config on the nodes, with no signal that they diverge, and root artifact verification cannot be applied while it holds.

## Expected behaviour

A temporary gate should be visibly temporary. Either the deploy surfaces that it is suppressing machine-config sync, or the two are decoupled so a Cilium rollout does not stop Talos reconciliation.

## Acceptance criteria

- A prod deploy that skips `cluster update` because the rollout gate is active **says so in its job summary** — not only as a skipped step.
- The suppression is **bounded**: after a defined threshold (deploy count or elapsed time) the deploy warns loudly, and the threshold is asserted by a test.
- A regression test fails if the gate can suppress `cluster update` with no surfaced signal.
- The decision is recorded: keep the coupling with a bound, or decouple autoscaler suspension from machine-config sync.
- Verified against a real deploy log, not only against the workflow source — the whole point here is that a correct-looking source produced a silent skip.

## Out of scope

Releasing the current Cilium rollout gate — that is an operational judgement about the rollout, not a repo change. The verify-block outcomes tracked in #2938 and #2922.

Rough size: small — a summary line plus a bounded threshold and its test. The decision on decoupling may be larger and can be a follow-up.

Part of #2627.

### PR DESCRIPTION
**Pull request by devantler**
> 🤖 Generated by the Agentic Engineer

## Why

The Cilium rollout gate skips the deploy pipeline's **only** Talos machine-config sync. That coupling
is deliberate and stays — what was not deliberate is that it happened silently, because a skipped
step inside a green deploy looks exactly like a step that ran and did nothing.

So machine config in Git quietly stopped being machine config on the nodes for **nine days** with no
signal anywhere, and it misdirected the diagnosis of two unrelated issues along the way.

## What

The deploy now says so in its job summary — what was skipped, since when, and what else is blocked
behind it — and warns loudly once the suppression outruns the rollout's intended one-week window.
Releasing the gate stays your operational call, so this never fails a deploy.

On today's state it fires immediately: 9 days elapsed against a 7-day bound.

The decision it records — keep the coupling, bound it — and the evidence are on #2951.

Fixes #2951

### CODE REVIEW

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

## Verification

**Programmatically tested.** CI 20 pass / 11 skipped / 0 pending at `ba72738c`.

Written test-first: the suite failed with `the gate-suppression reporter must be an executable
script` before any implementation existed. It now reports **17 assertions passed, 0 failed** — and
that count is read from the **runner's own log**, not just locally, because a step that is silently
`skipped` is exactly the class of defect this PR is about:

```
🧪 Validate Manifests  📣 Validate the rollout gate cannot suppress sync silently
passed: 17 failed: 0
```

**Six ablations, each firing with its own distinct message** — a green suite proves nothing until the
assertions are shown to bite:

| Ablation | Assertion that fired |
|---|---|
| activation marker removed | `gate is ACTIVE, so …rollout-gate-activated: <YYYY-MM-DD> must be declared` |
| `always()` removed from the step | `must run under always() so an earlier failure cannot hide the suppression` |
| reporter unwired from the deploy | `the deploy action must invoke the gate-suppression reporter` |
| threshold set to `999` | `the threshold must be a bound in (0,30] days` |
| warning condition disabled | `a suppression past the threshold must emit a ::warning:: annotation` |
| future-date guard neutered | `an activation date in the future must fail closed` |

The warning ablation is worth a note: my **first attempt was invalid** — the edit left a stray
argument line, so the suite failed on a shell error rather than on the missing warning, which looks
identical from the outside. Redone by flipping the condition to `if false;`, with `bash -n` and a
clean run confirming the script still worked before drawing any conclusion.

**Tried and evaluated as a user.** Run against the real repository state, this is what a deploy now
shows — it fires immediately, because the gate genuinely is overdue:

```
::warning::Talos machine-config sync has been suppressed by the Cilium rollout gate for 9 days
(threshold 7). Machine config in Git is diverging from the nodes, and platform#2922/#2938 are
blocked behind it. Finish or roll back the rollout — see the component runbook.
```

with a job-summary block naming the skipped sync, `Gate active since 2026-07-26`, `Elapsed 9 days`,
and what else is blocked behind it.

**Two self-review findings, fixed after the suite was already green:** a future activation date would
make elapsed negative so the bound could never trip — one typo would buy the rollout unlimited time
— and the marker read now takes the first declaration deterministically.

`always()` and `steps.<id>.outputs` are not assumed to work in a composite action: both idioms are
already used in this same `action.yml` (lines 234, 269, 272).

## What is deliberately not verified yet

The summary block appearing in a **real** deploy can only be observed after this merges and the next
merge-queue deploy runs. Acceptance criterion 5 asks for exactly that, so I will read the job summary
back and record it on #2951 rather than closing on green CI alone.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer


@coderabbitai review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer


@coderabbitai review

File: scripts/report-cilium-rollout-gate-suppression.sh [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  # head -1: the marker is prose-adjacent, so take the first declaration rather
+  # than letting a second mention of it anywhere in the file change the answer.
+  marker_line="$(grep -F "${activation_marker}" "${controllers_kustomization}" | head -1 || true)"
+  [[ -n "${marker_line}" ]] ||
+    fail "the rollout gate is active but ${activation_marker} <YYYY-MM-DD> is not declared beside the component reference"
+  # Parameter expansion, not sed: the marker contains '/', which would collide
+  # with sed's substitution delimiter.
+  activated="${marker_line##*"${activation_marker}"}"
+  activated="${activated#"${activated%%[![:space:]]*}"}"
+  activated="${activated%%[[:space:]]*}"
```
[the remark this answers is not in the document]
Comment (devantler) [reply scripts/report-cilium-rollout-gate-suppression.sh]: > 🤖 Generated by the Agentic Engineer

Valid — fixed in `5a33bb3`. The reporter now requires **exactly one** declaration and fails closed on 0 or >1, rather than resolving a duplicate to whichever came first. A stale marker left behind by a previous rollout was the realistic way this bites.

Test added with a real fixture (`PLATFORM_ROOT` pointed at a temp tree containing two markers) plus a single-marker control on the same path, so the assertion is about duplication and not about the fixture being unreadable. **Ablated**: reverting to `head -1` fires `a duplicated activation marker must fail closed, not silently take the first`.

File: scripts/report-cilium-rollout-gate-suppression.sh [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  printf 'cosign `verify` block (platform#2922, platform#2938).\n\n'
+  printf 'Release is an operational judgement about the Cilium rollout, not '
+  printf 'something a deploy decides: follow step 4–5 of the runbook in '
+  printf '`k8s/providers/hetzner/infrastructure/controllers/cilium/components/homogeneous-devices/kustomization.yaml`.\n'
+} >>"${summary}"
+
+if ((elapsed_days > warn_after_days)); then
+  printf '::warning::Talos machine-config sync has been suppressed by the Cilium rollout gate for %s days (threshold %s). Machine config in Git is diverging from the nodes, and platform#2922/#2938 are blocked behind it. Finish or roll back the rollout — see the component runbook.\n' \
+    "${elapsed_days}" "${warn_after_days}"
+fi
```
[the remark this answers is not in the document]
Comment (devantler) [reply scripts/report-cilium-rollout-gate-suppression.sh]: > 🤖 Generated by the Agentic Engineer

Valid — fixed in `5a33bb3`. Changed to `>=`. You are right that `warn_after_days=7` reads as "warn once this has run seven days", so warning on day eight silently made the real bound one day longer than the constant claims.

Two tests, deliberately paired: exactly-at-threshold **must** warn, and one day inside the bound **must not**. Without the second, a reporter that simply always warned would satisfy the first. **Ablated**: restoring `>` fires `the warning must fire ON the threshold day, not the day after`.

File: scripts/report-cilium-rollout-gate-suppression.sh [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  fail "could not parse the rollout gate activation date: '${activated}'"
+now_epoch="$(date -u +%s)"
+elapsed_days=$(((now_epoch - activated_epoch) / 86400))
+readonly activated_epoch now_epoch elapsed_days
+
+# A future activation date is a typo, and it would silently buy the rollout
+# unlimited extra time — the elapsed count goes negative and never trips the
+# bound. Fail closed rather than report a suppression as fresh forever.
+((elapsed_days >= 0)) ||
+  fail "the rollout gate activation date is in the future: '${activated}' (${elapsed_days} days)"
```
[the remark this answers is not in the document]
Comment (devantler) [reply scripts/report-cilium-rollout-gate-suppression.sh]: > 🤖 Generated by the Agentic Engineer

Valid, and the most interesting of the three — fixed in `5a33bb3`.

You are right about the mechanism: bash truncates integer division toward zero, so a date one day ahead produced `0`, which my `elapsed_days >= 0` guard read as "activated today". My own test only covered `+30d`, which yields `-30` and was caught — so the guard looked proven while the boundary case walked straight through it.

The guard now compares **epochs before any division**, and a next-day fixture is added.

**Investigating it turned up the underlying cause, which was broader than the guard.** The original code mixed two different anchorings: `now_epoch` came from `date -u +%s` (an instant) while the activation date was parsed date-only. That mixture is what let the difference be a sub-day negative in the first place. It also made the result platform-dependent — measured here:

```
now: 2026-08-04 23:29:56Z
BSD  date -j -f %Y-%m-%d 2026-08-05  =>  2026-08-05 23:29:56Z   # fills from the CURRENT time
GNU  date -d 2026-08-05              =>  2026-08-05 00:00:00Z   # midnight
```

So both sides now resolve through one function that anchors to **UTC midnight explicitly**, making the elapsed count an exact whole-day difference on GNU and BSD alike. A `1/3/9/40`-day probe asserts that exactness.

**Honest note on what is and is not demonstrated:** ablating the future guard fires the new test, so that guard is proven load-bearing. Reverting *only* the explicit midnight does **not** change observable behaviour now that both sides use the same parser — so I am recording that anchoring as robustness and clarity, not as a behaviour change I can show a failing test for.

Worth adding that this local reproduction was misleading in a way worth flagging: on macOS the next-day case yields exactly `-86400`, so the old guard *did* catch it here. Your finding is real on the Linux runner, where the delta is a sub-day negative. I would not have found it from the Mac.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer


@coderabbitai review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer


@codex review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer


CodeRabbit is rate-limited (45 min) and Codex is at its usage limit at this head, so requesting Bugbot. The bare trigger follows — Bugbot exact-matches the whole comment body, so it cannot carry this line.

**Comment by devantler** [comment]
@cursor review

**Review by devantler — COMMENTED** [review]
> 🤖 Generated by the Agentic Engineer

## Self-review (fallback — CodeRabbit, Codex and Cursor Bugbot unavailable)

**Reviewed commit:** `5a33bb35c8d44cf12134b82d6af80af0e0482880`

### Per-lane state, checked directly on this PR at this head

| Lane | Evidence | Time |
|---|---|---|
| CodeRabbit | ack `⚠️ Action not completed / Review rate limited`; summary `Review limit reached`, **45 minutes** | 23:38:11Z |
| Codex | `You have reached your Codex usage limits for code reviews` — no window, needs credits | 23:39:14Z |
| Cursor Bugbot | check-run `conclusion=neutral` + `output.title=Error` (run never started); `cursor[bot]`: usage limit, admin must raise | 23:40:04Z |

### Important context — this change is not unreviewed

CodeRabbit **did** review this PR substantively at `0d81b691` and raised three 🟠 Major findings. All three were valid, all three are fixed in `5a33bb3` with fixtures and ablations, and all three threads are replied to and resolved. The delta this fallback covers is exactly those fixes.

### What I checked

- **The three fixes do what the findings asked.** Exactly one activation marker is now required (0 or >1 fails closed); the future-date guard compares epochs *before* dividing, so a next-day date can no longer truncate to `0`; the warning fires on the threshold day via `>=`.
- **Ablations, each with a syntax + clean-run precondition so a broken mutation cannot masquerade as a catch.** Neutralising the future guard fires the next-day assertion; reverting `>=` fires the exact-threshold assertion; restoring `head -1` fires the duplicate-marker assertion.
- **Honest limit:** reverting *only* the explicit UTC-midnight anchoring changes no observable behaviour now that both sides resolve through one parser, so I record that as robustness and clarity — not as something a failing test demonstrates.
- **Fail-closed reachability.** `fail()` exits non-zero, which would fail the deploy step. That path is only reachable via a malformed or duplicated marker, which the CI contract test rejects first, so it is defence in depth rather than a new way for a deploy to break. An overdue gate warns and still exits 0 — releasing it stays an operational judgement.
- **Blast radius.** Additive: one new step, one new script, one new test, two CI wiring entries, one comment marker. No existing step's condition or ordering changed; the three pre-existing Cilium contract tests still pass.
- `always()` and `steps.<id>.outputs` are not assumed to work in a composite action — both idioms already appear in this same `action.yml`.

### Not verified here, deliberately

The job-summary block rendering in a **real** deploy. That is acceptance criterion 5 on #2951 and can only be observed post-merge; I will read the next merge-queue deploy's summary back and record it on the issue rather than closing on green CI.

**Verdict: no P0/P1 findings.**
